### PR TITLE
report error to PD server

### PR DESCRIPTION
**Pull request by YuJuncen**
(This PR is based on works of #41)

This PR added some error reporting method for PD, and would trigger the pause-and-report if flush failed too many times.

### CODE REVIEW

File: components/backup-stream/src/endpoint.rs [hunk 1]
```diff
@@ -208,10 +221,33 @@ where
...
+        self.pool.block_on(async move {
+            // TODO: also pause the task using the meta client.
+            let mut last_error = StreamBackupError::new();
+            last_error.set_error_code(err.error_code().code.to_owned());
+            last_error.set_error_message(err.to_string());
+            last_error.set_store_id(store_id);
+            last_error.set_happen_at(TimeStamp::physical_now());
+            if let Err(err_report) = meta_cli.report_last_error(&task, last_error).await {
+                err_report.report(format_args!("failed to upload error {}", err_report));
+                // Let's retry reporting after 5s.
```
Comment (3pointer) [inline components/backup-stream/src/endpoint.rs:242]: why keep reporting this error every 5s?🤔

Comment (YuJuncen) [reply components/backup-stream/src/endpoint.rs:242]: If failed to report, we would retry report error by this. (Just a rough retry implement)

**Review by 3pointer — APPROVED** [review]
LGTM

File: components/backup-stream/src/router.rs [hunk 2]
```diff
@@ -451,14 +453,22 @@ impl RouterInner {
...
                 // set false to flushing whether success or fail
                 task_info.set_flushing_status(false);
                 task_info.update_flush_time();
+
+                if let Err(e) = result {
+                    e.report("failed to flush task.");
+                    warn!("backup steam do flush fail"; "err" => ?e);
+                    if task_info.flush_failure_count() > FLUSH_FAILURE_BECOME_FATAL_THRESHOLD {
+                        // NOTE: Maybe we'd better record all errors and send them to the client?
+                        try_send!(
```
Comment (joccau) [inline components/backup-stream/src/router.rs:465]: Pretty Good!

File: components/backup-stream/src/endpoint.rs [hunk 3]
```diff
@@ -212,10 +214,33 @@ where
...
-    fn on_fatal_error(&self, _task: String, _err: Box<Error>) {
-        // This is a stub.
-        // TODO: implement the feature of reporting fatal error to the meta storage,
-        //       and pause the task then.
+    fn get_meta_client(&self) -> MetadataClient<S> {
+        self.meta_client.as_ref().unwrap().clone()
+    }
+
+    fn on_fatal_error(&self, task: String, err: Box<Error>) {
+        // Let's pause the task locally first.
```
Comment (joccau) [inline components/backup-stream/src/endpoint.rs:222]: We will need another PR about `meta_client.pause(taskname)` later.

**Review by joccau — APPROVED** [review]
rest LGTM

File: components/backup-stream/src/endpoint.rs (hunk 3 above)
Comment (YuJuncen) [reply components/backup-stream/src/endpoint.rs:222]: done

**Pull request merged**